packages/theme/src/cli/utilities/theme-uploader.ts

```diff
@@ -139,21 +140,32 @@ function buildDeleteJob(
   const remoteFilesToBeDeleted = getRemoteFilesToBeDeleted(remoteChecksums, themeFileSystem)
   const orderedFiles = orderFilesToBeDeleted(remoteFilesToBeDeleted)
 
-  let failedDeleteAttempts = 0
   const progress = {current: 0, total: orderedFiles.length}
-  const promise = Promise.all(
-    orderedFiles.map((file) =>
-      deleteThemeAsset(theme.id, file.key, session)
-        .catch((error: Error) => {
-          failedDeleteAttempts++
-          if (failedDeleteAttempts > 3) throw error
-          createSyncingCatchError(file.key, 'delete')(error)
-        })
-        .finally(() => {
-          progress.current++
-        }),
-    ),
-  ).then(() => {
+  if (orderedFiles.length === 0) {
+    return {progress, promise: Promise.resolve()}
+  }
+
+  const deleteBatches = []
+  for (let i = 0; i < orderedFiles.length; i += MAX_BATCH_FILE_COUNT) {
```

Comment:
I think it's a mistake to create promises for all files immediately, so long as throttling doesn't exist. I think you're likely to end up with rate limiting errors, timeouts, etc. when making sweeping changes to large themes (or really anything with an automated build process resulting in lots of updates).

I also think it's a mistake to keep this a synchronous function. `async`/`await` exists for a reason, and this code is getting quite complex with nothing to catch failure, so it's possible things could just never resolve (e.g. the HTTP request totally fails and throws an error) rather than throwing an error that gets displayed to the user.

I won't hold up review as this isn't my domain of expertise, but I did want to make sure you're aware of these concerns.